### PR TITLE
[WIP]  Add parsing best practices doc

### PR DESCRIPTION
**Pull request by VibhuJawa**
This is the parsing practices best practices doc to link to our regex blog . 
 
This will also close https://github.com/rapidsai/custrings/issues/225 .

I am not sure if we are still in freeze or not but wanted to get the initial draft reviewed, will re target if needed. 

The rendered doc looks like below:

image

### CODE REVIEW

File: docs/source/parsing_best_practices.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+### Sorting by length:
+
+If your data has unequal length distribution (For example, the longest string might be 1000x longer than the shortest string) and you are doing heavy regex heavy operations, sorting the strings can help to improve the performance.
+
+##### Why:
+
+As regexes operate one character at a time, longer length strings take longer to evaluate and operating on strings in parallel improves throughput.   The GPU works most efficiently when adjacent threads in a warp work in lock step. Wildly different string lengths within adjacent threads means that the threads operating on shorter strings become idle, reducing utilization.
+
+Sorting helps to obtain better wrapping helping adjacent threads in a warp work in lock step to improve utilization .
```
Comment (davidwendt) [inline docs/source/parsing_best_practices.md]: Better 'wrapping'?  Assume you mean better 'warping' though I don't think that is real term. How about this:
_Sorting helps to improve performance by helping adjacent threads in a warp work in lock step to improve utilization_

File: docs/source/parsing_best_practices.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+Lets walk throw a simple example to see the performance differences for ourselves. 
+
+Lets take a simple example where we have to remove `-` character between 2 alphabetical characters while keeping the `-` between numerical characters on a dataset with 4 million short strings and 50k long strings . 
+
+
+```python
+import random
+import cudf
+
+small_string = "Exmaple-string 123-456-789"
```
Comment (davidwendt) [inline docs/source/parsing_best_practices.md]: Exmaple is a tree that used to be maple. I think you mean _Example_ :-)

File: docs/source/parsing_best_practices.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+
+
+### Using `nvtext`
+
+`nvtext` provides out-of-the-box functions which make these computations like removing multiple characters and stop words much faster.
+
+
+```python
+%time cudf_sr = cudf_sr.str.replace_multi(filters, ' ',regex=False)
```
Comment (davidwendt) [inline docs/source/parsing_best_practices.md]: The `replace_multi` is an `nvstrings` method and not an `nvtext` method. The `nvtext` method is `replace_tokens` and uses delimiters to locate tokens. The `nvstrings` `replace_multi` replaces strings and not tokens.

File: docs/source/parsing_best_practices.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
+# nvstrings Parsing Best Practices
+
+
+## Regex Performence Tips:
```
Comment (davidwendt) [inline docs/source/parsing_best_practices.md]: Performance is misspelled twice that I can find.

File: docs/source/parsing_best_practices.md (hunk 3 above)
Comment (VibhuJawa) [reply docs/source/parsing_best_practices.md]: Yup, this was incorrect, this should just have been `nvstrings`.replace_multi .

For `nvtext.replace_tokens`, I  can create another example , where i remove `stop-words` and show different regex  performance for that . 

https://rapidsai.github.io/projects/nvstrings/en/0.9.0/api.html?source=post_page---------------------------#nvtext.replace_tokens

File: docs/source/parsing_best_practices.md (hunk 1 above)
Comment (VibhuJawa) [reply docs/source/parsing_best_practices.md]: Should be fixed after commit: https://github.com/rapidsai/custrings/pull/397/commits/8f5ce58efd95160ba4ff0f2d2014f494da1adb34

File: docs/source/parsing_best_practices.md (hunk 2 above)
Comment (VibhuJawa) [reply docs/source/parsing_best_practices.md]: Should be fixed now.  https://github.com/rapidsai/custrings/pull/397/commits/8f5ce58efd95160ba4ff0f2d2014f494da1adb34

File: docs/source/parsing_best_practices.md (hunk 3 above)
Comment (VibhuJawa) [reply docs/source/parsing_best_practices.md]: Changed this to add different  examples for both,  `nvstrings.replace_multi` and `nvtext.replace_tokens`

File: docs/source/parsing_best_practices.md (hunk 4 above)
Comment (VibhuJawa) [reply docs/source/parsing_best_practices.md]: Did a spell check again and fixed this.  Hopefully, there are no more typo's .

**Pull request closed without merging**